k9mail/src/main/java/com/fsck/k9/activity/setup/OpenPgpAppSelectDialog.java

```diff
@@ -0,0 +1,294 @@
+ * limitations under the License.
+ */
+
+package com.fsck.k9.activity.setup;
+
+
+import java.util.ArrayList;
+import java.util.List;
+
+import android.app.Activity;
+import android.app.AlertDialog;
+import android.app.Dialog;
+import android.app.DialogFragment;
+import android.content.Context;
+import android.content.DialogInterface;
+import android.content.Intent;
+import android.content.pm.ResolveInfo;
+import android.graphics.drawable.Drawable;
+import android.net.Uri;
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.ArrayAdapter;
+import android.widget.ListAdapter;
+import android.widget.TextView;
+
+import com.fsck.k9.K9;
+import com.fsck.k9.Preferences;
+import com.fsck.k9.R;
+import com.fsck.k9.preferences.StorageEditor;
+import com.fsck.k9.ui.dialog.ApgDeprecationWarningDialog;
+import org.openintents.openpgp.util.OpenPgpApi;
+import org.openintents.openpgp.util.OpenPgpAppPreference;
+
+
+/**
+ * Does not extend ListPreference, but is very similar to it!
+ * http://grepcode.com/file_/repository.grepcode.com/java/ext/com.google.android/android/4.4_r1/android/preference/ListPreference.java/?v=source
+ */
+public class OpenPgpAppSelectDialog extends Activity {
+    private static final String OPENKEYCHAIN_PACKAGE = "org.sufficientlysecure.keychain";
+    private static final String APG_PROVIDER_PLACEHOLDER = "apg-placeholder";
+    private static final String PACKAGE_NAME_APG = "org.thialfihar.android.apg";
+
+    private static final String MARKET_INTENT_URI_BASE = "market://details?id=%s";
+    private static final Intent MARKET_INTENT = new Intent(Intent.ACTION_VIEW, Uri.parse(
+            String.format(MARKET_INTENT_URI_BASE, OPENKEYCHAIN_PACKAGE)));
+
+    private static final ArrayList<String> PROVIDER_BLACKLIST = new ArrayList<>();
+
+    static {
+        // Unfortunately, the current released version of APG includes a broken version of the API
+        PROVIDER_BLACKLIST.add(PACKAGE_NAME_APG);
+    }
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setTheme(K9.getK9Theme() == K9.Theme.LIGHT ?
+                R.style.Theme_K9_Dialog_Translucent_Light : R.style.Theme_K9_Dialog_Translucent_Dark);
+
+        if (savedInstanceState == null) {
+            showOpenPgpSelectDialogFragment();
+        }
+    }
+
+    private void showOpenPgpSelectDialogFragment() {
+        OpenPgpAppSelectFragment fragment = new OpenPgpAppSelectFragment();
+        fragment.show(getFragmentManager(), "openpgp_select");
+    }
+
+    private void showApgDeprecationDialogFragment() {
+        ApgDeprecationDialogFragment fragment = new ApgDeprecationDialogFragment();
+        fragment.show(getFragmentManager(), "apg_deprecation");
+    }
+
+    public static class OpenPgpAppSelectFragment extends DialogFragment {
```

Comment:
You're right, it's mostly because this new code lives in k-9 where the old one was in the library. I didn't want to rewrite too much of the lib for no reason at this point, but we could move to using the activity as well from the preference later on.